File: day01/hs/Main.hs
#!/usr/bin/env runhaskell
import           Data.Char     (isDigit)
import           Data.Function ((&))
import           Data.Maybe    (maybeToList)

convertLeadingDigit :: String -> Maybe Char
convertLeadingDigit ('o':'n':'e':_)         = Just '1'
convertLeadingDigit ('t':'w':'o':_)         = Just '2'
convertLeadingDigit ('t':'h':'r':'e':'e':_) = Just '3'
convertLeadingDigit ('f':'o':'u':'r':_)     = Just '4'
convertLeadingDigit ('f':'i':'v':'e':_)     = Just '5'
convertLeadingDigit ('s':'i':'x':_)         = Just '6'
convertLeadingDigit ('s':'e':'v':'e':'n':_) = Just '7'
convertLeadingDigit ('e':'i':'g':'h':'t':_) = Just '8'
convertLeadingDigit ('n':'i':'n':'e':_)     = Just '9'
convertLeadingDigit (x:_) | isDigit x       = Just x
convertLeadingDigit _                       = Nothing

digitsOnly :: String -> String
digitsOnly []     = []
digitsOnly (x:xs) = maybeToList (convertLeadingDigit (x:xs)) ++ digitsOnly xs

main = do
  ls <- lines <$> readFile "input"
  putStr "Part 1: "
  print $ solve1 ls
  putStr "Part 2: "
  print $ solve2 ls

solve1 :: [String] -> Int
solve1 ls = sum . map read $ digits
  where
    digits = map (filter isDigit) ls
           & map (\xs -> [head xs, last xs])

solve2 :: [String] -> Int
solve2 ls = sum . map read $ vals
  where
    vals = map digitsOnly ls
         & map (\xs -> [head xs, last xs])

File: day01/rs/src/main.rs
use std::fs;

fn main() {
    let cont = fs::read_to_string("input").unwrap();
    let lines: Vec<_> = cont.lines().collect();
    println!("Part 1: {}", solve1(&lines));
    println!("Part 2: {}", solve2(&lines));
}

fn solve1<'a>(lines: &'a Vec<&str>) -> i32 {
    lines
        .iter()
        .map(|ln| {
            let mut digits = ln.chars().filter(|x| x.is_digit(10)).peekable();
            let first = digits.peek().unwrap().clone();
            let last = digits.last().unwrap();

            (first.to_digit(10).unwrap() * 10 + last.to_digit(10).unwrap()) as i32
            // * Initially wrote it like this, so I'm keeping it
            // [first.to_owned(), last.to_owned()]
            //     .iter()
            //     .collect::<String>()
            //     .parse::<i32>()
            //     .unwrap()
        })
        .sum()
}

const NUMBERS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn solve2<'a>(lines: &'a Vec<&str>) -> i32 {
    lines
        .iter()
        .map(|ln| {
            let mut digits = ln.chars().enumerate().filter_map(|(i, c)| {
                if c.is_digit(10) {
                    return c.to_digit(10).map(|x| x as i32);
                }

                NUMBERS
                    .iter()
                    .position(|x| ln[i..].starts_with(x))
                    .map(|x| (x + 1) as i32)
                // * Initially wrote it like this, so I'm keeping it
                // if let Some(idx) = DIGITS.iter().position(|x| ln[i..].starts_with(x)) {
                //     Some((idx + 1) as i32)
                // } else {
                //     None
                // }
            }).peekable();

            digits.peek().unwrap() * 10 + digits.last().unwrap()
        })
        .sum::<i32>()
}
